### ISSUE TITLE
How to change all lines of the selection? (code-folding safe)

### ISSUE THREAD

**Issue by EricGrange** [bonecode/TBCEditor#73]
What is the best way to change multiple lines in a code-folding safe way?

The purpose is to make a utility that operates on all lines of the selection, for instance to comment out a block of code with '//'

Modifying the selection and using SelectedText has a side effect on bookmarks (moves all the bookmarks of the selection down).

Lines[] only exposed non-collapsed lines.

CreateUncollapsedLines is a copy, and thus basically read-only.

**Comment by bonecode** [comment]
SelectedText it is. Then there is a bug with bookmarks. I think I know what it is all about because I was fixing it lately. :)

**Comment by EricGrange** [comment]
I am not sure SelectedText has a "correct" way of preserving bookmarks, because you cannot know what the text really has been replaced with. I tested a few other editors, and they all have the same behavior as TBCEditor currently has.

Another "selection block" operation would f.i. be a "surround with" refactor, in which the number of lines change, in that case, preserving the original bookmark position would not make sense. However if the surround with was comprised of "insertline" and "replaceline" commands, bookmarks could be preserved intelligently (by "bookmarks" I also have in mind things like breakpoints)

In the same way there are other refactors that need to work in terms of source code lines, and may not have a direct relationship to the selection, for instance a rename refactor (where the selection is only used to pinpoint the symbol to rename, but you rename by changing many lines in the current editor), and all the more complex refactors or code wizards.

So having a way to query and update the "actual" source lines directly is required, as is having a way to insert and delete actual source lines.
